src/support/utils.js

```diff
@@ -175,117 +173,131 @@ export async function getRUMDomainkey(baseURL, context) {
- *
- * This function takes a URL as input and processes its pathname to extract
- * keywords. Each segment of the path is treated as a keyword, and segments
- * are ranked based on their position in the path. The keyword closer to the
- * end of the path has a higher rank. File extensions, if present, are removed
- * from the last segment.
- *
- * @param {string} url - The URL from which to extract keywords.
- * @param {Object} log - The logger object for logging messages.
- * @returns {Array<{keyword: string, rank: number}>} An array of objects, each containing
- * the keyword and its rank. The rank is determined by the position of the segment
- * in the URL path, with higher ranks for segments closer to the end.
- *
- * @example
- * // Returns [{ keyword: 'foo', rank: 2 }, { keyword: 'bar', rank: 1 }]
- * extractKeywordsFromUrl('http://www.example.com/foo/bar');
- *
- * @example
- * // Returns [{ keyword: 'foo bar', rank: 2 }, { keyword: 'baz', rank: 1 }]
- * extractKeywordsFromUrl('http://www.example.com/foo-bar/baz.html');
- */
-export const extractKeywordsFromUrl = (url, log) => {
+const getFileContentFromS3 = async (s3Client, bucket, key, log) => {
+  const getCommand = new GetObjectCommand({
+    Bucket: bucket,
+    Key: key,
+  });
+  const data = await s3Client.send(getCommand);
   try {
-    const urlObjc = new URL(url);
-    const path = urlObjc.pathname;
-
-    const segments = path.split('/').filter((segment) => segment.length > 0);
+    const content = await data.Body.transformToString();
+    return JSON.parse(content);
+  } catch (error) {
+    log.error('Error parsing S3 object:', error);
+    return null;
+  }
+};
 
-    // Remove file extensions from the last segment if present
-    if (segments.length > 0) {
-      const lastSegment = segments[segments.length - 1];
-      segments[segments.length - 1] = lastSegment.replace(/\.[^/.]+$/, '');
-    }
+const extractScrapedMetadataFromJson = (data, log) => {
+  try {
+    log.debug(`Extracting data from JSON (${data.finalUrl}:`, JSON.stringify(data.scrapeResult.tags));
+    const finalUrl = data.finalUrl || '';
+    const title = data.scrapeResult.tags?.title || '';
+    const description = data.scrapeResult.tags?.description || '';
+    const h1Tags = data.scrapeResult.tags?.h1 || [];
+    const h1Tag = h1Tags.length > 0 ? h1Tags[0] : '';
 
-    // Map segments to an array of objects with segment and rank
-    return segments.map((segment, index) => ({
-      keyword: segment.replace(/-/g, ' '),
-      rank: segments.length - index, // Rank: higher for segments closer to the end
-    }));
+    return {
+      url: finalUrl,
+      title,
+      description,
+      h1: h1Tag,
+    };
   } catch (error) {
-    log.error('Invalid URL:', error);
+    log.error('Error extracting data:', error);
+    return null;
+  }
+};
+
+const extractLinksFromHeader = (data, baseUrl) => {
+  if (!data || !data.scrapeResult || !data.scrapeResult.rawBody) {
```

Comment:
could use `isNonEmptyObject(data?.scrapeResult) && hasText(data.scrapeResult.rawBody)`

Reply:
does this warrant a log info or warn?